rfcs/0001-soxin-design.md

```diff
@@ -0,0 +1,92 @@
+---
+feature: soxin design
+start-date: 2020-05-21
+author:
+  - Marc 'risson' Schmitt
+  - Wael Nasreddine
+related-issues: None
+---
+
+# Summary
+[summary]: #summary
+
+This RFC describes how the Soxin project will be structured. The goal is to
+avoid past mistakes made in trying to create a similar project. The attitude
+we are trying to adopt is not assuming anything about a user's background,
+workflow, preferences, all while keeping a sane set of proposed default
+settings.
+
+# Motivation
+[motivation]: #motivation
+
+Soxin's goal is to provide an easy way for newcomers to use Nix and related
+technologies, such as NixOS, home-manager and nix-darwin. This will include,
+but is not limited to, default configuration for widely used programs, a start
+guide on how to get started and hopefully an installer.
+
+# Detailed design
+[design]: #detailed-design
+
+To achieve all those goals and stay in line with the core principles we set,
+a clear design is necessary.
+
+## Global settings
+
+Some settings are meant to be set globally for a system, such as keyboard
+layout(s), theme, fonts and programming languages used. As such, Soxin will
+provide a way to set those under the `soxin.settings` module. Those settings
+will be used by all programs and services that require them. However, some
+users might want to override those settings on a per-program basis. As such,
+Soxin must provide a library function that is used to create a module, with
+the same global settings but only for that module. Those module-scoped settings
+will default to the global ones. For instance, this library function could look
+like:
+
+```
+mkSoxinModule = { config, name, extraOptions ? { } }: lib.recursiveUpdate {
+  enable = mkEnableOption name;
+
+  keyboardLayout = mkOption {
+    type = layoutModule; # see the keyboard module for an explanation of this
+    default = config.soxin.settings.keyboard.defaultLayout;
+    description = "Keyboard layout to use for ${name}";
+  };
+} extraOptions;
+```
+
+and be used like:
+
+```
+options = {
+  soxin.programs.neovim = mkSoxinModule {
```

Comment:
Isn't this the same thing as doing

```
soxin.programs.neovim = options.soxin.settings // {
  /* ... */
};
```

provided options were received at the top of the module.

Reply:
As presented yes, but it could be a lot more modulable, if we had feature switches like this:
```
soxin.programs.neovim = mkSoxinModule { keyboard = false; };
```
If we don't want the keyboard to be configurable for this module. An example of that would be the i3 module, for which we will only configure the theme, and leave it to the user to configure their keybindings.